nextstrain_profiles/nextstrain-gisaid/builds.yaml

```diff
@@ -215,6 +217,15 @@ subsampling:
       max_sequences: 500
       sampling_scheme: "--probabilistic-sampling"
       exclude: "--exclude-where 'region={region}'"
+  
+  # Custom subsampling logic for group by clade
+  nextstrain_clades:
+    clades:
+      group_by: "Nextstrain_clade"
+      max_sequences: 300
+    clades_month: # To ensure at least one sample from the earliest month
+      group_by: "Nextstrain_clade year month"
+      max_sequences: 200
```

Comment:
I'm a bit confused, is this supposed to be 2 subsamplings of 300 and 200 for a total of 500 sequences? There are only 318 in the staging dataset that John linked https://nextstrain.org/staging/ncov/open/trial/reference_targets/reference

Reply:
I could see how separate-sampling means there're 2 passes through the dataset (slight inefficiency). I'm open to ways to speed this up. 

* If I only do `clades_month`, then clades with a short duration will almost disappear in the reference tree. 
* If I only do `clades`, then I risk not capturing an early strain. 

Ideally, I'd be listing a set of strains (first instance in a clade) to force-include

Reply:
Ah yeah there are some limitations here, which can be solved with a database approach in augur.

### 1

> separate-sampling means there're 2 passes through the dataset

The bottleneck is loading all the data for every call to `augur filter`. The ideal state is to load the data once then perform queries on it.

### 2

> Ideally, I'd be listing a set of strains (first instance in a clade) to force-include

This isn't possible with the current interface for `augur filter`. In pandas terms, we would need something like:

```py
df.sort_values(['year', 'month'], ascending=True).groupby(['Nextstrain_clade', 'year', 'month']).first()
```

The closest interface available in `augur filter` is `--query` as an alias to [`pandas.DataFrame.query`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.query.html), which doesn't support this.  A global `groupby` is actually out of the picture now that metadata is read in chunks.

In SQLite, this would be something like:

```sql
WITH grouped_rows AS (
    SELECT
        strain,
        ROW_NUMBER() OVER (
            PARTITION BY Nextstrain_clade, year, month
            ORDER BY year ASC, month ASC
        ) AS group_i
    FROM metadata  -- need to join with extracted year/month
)
SELECT strain
FROM grouped_rows
WHERE group_i = 1
```

Reply:
(No action needed in this PR, you've answered my question!)

Reply:
> My understanding was that the two samples were combined and deduplicated

Correct, but I'm surprised that there would be so many duplicates -- circa 182/200 of the "clades_month" strains are also in the "clades" if I'm understanding correctly.  The logs should report how many strains were picked for each filter ("sample") - can you double check these are (close to) 200 and 300?

Reply:
Sure thing!

```
snakemake -pf auspice/ncov_open_reference.json \
  --profile nextstrain_profiles/nextstrain-open \
  &> output_capture.txt
```

Checking the stdout/stderr, I can see `clades` sampling:

```
augur filter \
  --metadata nextstrain-data/files/ncov/open/metadata.tsv.gz \
  --include defaults/include.txt \
  --exclude defaults/exclude.txt \
  --group-by Nextstrain_clade \
  --subsample-max-sequences 300 \
  --output-strains results/reference/sample-clades.txt \
  2>&1 | \
  tee logs/subsample_reference_clades.txt
```

and `clades_month` sampling:

```
augur filter \
  --metadata nextstrain-data/files/ncov/open/metadata.tsv.gz \
  --include defaults/include.txt \
  --exclude defaults/exclude.txt \
  --group-by Nextstrain_clade year month \
  --subsample-max-sequences 200 \
  --output-strains results/reference/sample-clades_month.txt \
  2>&1 | \
  tee logs/subsample_reference_clades_month.txt
```

The `clades` sampling returned 262 strains:

```
less logs/subsample_reference_clades.txt
Sampling at 10 per group.
3851953 strains were dropped during filtering
        562 of these were dropped because they were in defaults/exclude.txt
        2 strains were added back because they were in defaults/include.txt
        3851393 of these were dropped because of subsampling criteria
262 strains passed all filters
```

The `clades_month` sampling returned 202 strains:

```
WARNING: Asked to provide at most 200 sequences, but there are 422 groups.
Sampling probabilistically at 0.4761 sequences per group, meaning it is possible to have more than the requested maximum of 200 sequences after filtering.
3852013 strains were dropped during filtering
        562 of these were dropped because they were in defaults/exclude.txt
        707 were dropped during grouping due to ambiguous year information
        22270 were dropped during grouping due to ambiguous month information
        2 strains were added back because they were in defaults/include.txt
        3828476 of these were dropped because of subsampling criteria
202 strains passed all filters
```

Since there's a warning of `there are 422` groups, I'm also happy to increase the number of `clades_month` to 450. What do you think?

Reply:
Although my local run seems to have more samples (459 unique strains), hmm...

```
cat results/reference/sample-clades_month.txt \
    results/reference/sample-clades.txt  |\
  head -n2
#> Wuhan/WH01/2019
#> Wuhan-Hu-1/2019
#> LR991346

cat results/reference/sample-clades_month.txt \
    results/reference/sample-clades.txt  |\
  sort | uniq | wc -l
#>      459
```

But then I looked at the fasta files which does eventually filter samples down to 348 strains.

```
ls -ltr results/reference/*.fasta
#> 13M Feb 23 11:11 results/reference/aligned.fasta
#> 13M Feb 23 11:11 results/reference/masked.fasta
#> 11M Feb 23 11:11 results/reference/filtered.fasta
#> 11M Feb 23 11:11 results/reference/masked_filtered.fasta

grep -c ">" results/reference/*.fasta
#> results/reference/aligned.fasta:439
#> results/reference/filtered.fasta:348
#> results/reference/masked.fasta:439
#> results/reference/masked_filtered.fasta:348   #<= this one
```

Reply:
I found the `augur filter` command with `snakemake --dry-run`:

```sh
augur filter \
--sequences results/reference/masked.fasta \
--metadata results/reference/metadata_with_index.tsv \
--include defaults/include.txt \
--query "(_length >= 27000)" \
--max-date 2022-02-24 \
--min-date 2019.74 \
--exclude-ambiguous-dates-by any \
--exclude defaults/exclude.txt results/reference/excluded_by_diagnostics.txt \
--exclude-where division='USA'
--output results/reference/filtered.fasta \
--output-log results/reference/filtered_log.tsv
```

and this was the output:

```
113 strains were dropped during filtering
	65 of these were dropped because they were in results/reference/excluded_by_diagnostics.txt
	24 of these were dropped because of 'division=USA'
	17 of these were filtered out by the query: "(_length >= 27000)"
	7 of these were dropped because of their ambiguous date in any
	2 strains were added back because they were in defaults/include.txt
323 strains passed all filters
```

Input to that has 436 strains (1st row is header):

```sh
wc -l results/reference/metadata_with_index.tsv
# 437
```

So the end result is 436 - 113 = 323